actors/evm/src/interpreter/precompiles.rs

```diff
@@ -57,79 +78,145 @@ impl<BS: Blockstore, RT: Runtime<BS>> Precompiles<BS, RT> {
+    Slice(&'a [u8]),
+    Vec(Vec<u8>),
 }
+impl<'a> Data<'a> {
+    fn slice(&self) -> &[u8] {
+        match self {
+            Self::Slice(s) => s,
+            Self::Vec(v) => v,
+        }
+    }
+
+    /// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/common/bytes.go#L108
+    fn read_right_pad(input: &'a [u8], len: usize) -> Self {
+        if len <= input.len() {
+            Self::Slice(input)
+        } else {
+            let mut padded = vec![0u8; len];
+            padded[..input.len()].copy_from_slice(input);
+            Self::Vec(padded)
+        }
+    }
+
+    /// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/core/vm/common.go#L54
+    fn get_data(input: &'a [u8], start: usize, len: usize) -> Self {
+        let start = start.min(input.len());
+        let end = (start + len).min(input.len());
+        Self::read_right_pad(&input[start..end], len)
+    }
+
+    /// read 32 bytes (u256) from buffer or pad
+    fn read_u256_padded(input: &[u8], start: usize) -> U256 {
+        let input = Data::get_data(input, start, 32);
+        U256::from_big_endian(input.slice())
+    }
 
-/// read 32 bytes (u256) from buffer, pass in exit reason that is desired
-/// returns 0 if failed to read
-fn read_u256_infalliable(buf: &[u8], start: usize) -> U256 {
-    let slice = buf.get(start..start + 32).unwrap_or(&[0u8; 32]);
-    U256::from_big_endian(slice)
+    fn read_bigint(input: &'a [u8], start: usize, len: usize) -> BigUint {
+        let data = Self::get_data(input, start, len);
+        BigUint::from_bytes_be(data.slice())
+    }
 }
 
+// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/core/vm/contracts.go#L165
+/// recover a secp256k1 pubkey from a hash, recovery byte, and a signature
 fn ec_recover<RT: Primitives>(rt: &RT, input: &[u8]) -> PrecompileResult {
-    let mut buf = [0u8; 128];
-    buf[..input.len().min(128)].copy_from_slice(&input[..input.len().min(128)]);
+    let data = Data::read_right_pad(input, 128);
+    let input = data.slice();
 
-    let mut hash = [0u8; SECP_SIG_MESSAGE_HASH_SIZE];
-    let mut sig = [0u8; SECP_SIG_LEN];
+    let hash: &[u8; SECP_SIG_MESSAGE_HASH_SIZE] = input[0..32].try_into().unwrap();
+    let r = BigUint::from_bytes_be(&input[64..96]);
+    let s = BigUint::from_bytes_be(&input[96..128]);
+    let recovery_byte = input[63];
+
+    // recovery byte is a single byte value but is represented with 32 bytes, sad
+    let v = if input[32..63] == [0u8; 31] && matches!(recovery_byte, 27 | 28) {
+        recovery_byte - 27
+    } else {
+        return Ok(Vec::new());
+    };
 
-    hash.copy_from_slice(&input[..32]);
-    sig.copy_from_slice(&input[64..]); // TODO this assumes input is exactly 65 bytes which would panic if incorrect
+    let valid = if r <= BigUint::one() || s <= BigUint::one() {
+        false
+    } else {
+        let secp256k1_n = BigUint::parse_bytes(
```

Comment:
Can we avoid parsing here?

Reply:
lazy-static or crypto-bigint which has const init functions

Reply:
implemetned with lazy-static for now